bip-internalkey.md

```diff
@@ -0,0 +1,70 @@
+```
+BIP: XXX
+Layer: Consensus (soft fork)
+Title: OP_INTERNALKEY
+Author: Brandon Black <freedom@reardencode.com>, Jeremy Rubin <j@rubin.io>
+Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-XXXX
+Status: Draft
+Type: Standards Track
+Created: 2023-12-22
+License: BSD-3-Clause
+```
+
+## Abstract
+
+This BIP describes a new tapscript opcode (`OP_INTERNALKEY`) which
+pushes the taproot internal key to the stack.
+
+## Specification
+
+When verifying taproot script spends having leaf version `0xc0` (as defined in
+[BIP 342]), `OP_INTERNALKEY` replaces `OP_SUCCESS203` (0xcb). `OP_INTERNALKEY`
+pushes the taproot internal key, as defined in [BIP 341], to the stack.
```

Comment:
It seems to me that this should be perhaps a bit more specific what exactly is being pushed to the stack. I assume this is referring to 

> Let p = c[1:33] and let P = lift_x(int(p)) where lift_x and [:] are defined as in [BIP340](https://bips.xyz/0340#design). Fail if this point is not on the curve. […] q is referred to as taproot output key and p as taproot internal key.

but it would not hurt to clarify.

Reply:
"the _taproot internal key_ denoted as _P_ in [BIP 341]"?

Reply:
My point is that the "taproot internal key" refers to an abstract object in BIP 341

> "q is referred to as taproot output key and p as taproot internal key."

and it wouldn’t hurt to simply describe what you are referring to, instead of sending people to parse another BIP to get clued in.

How about:

> "pushes the 32-byte x-only representation _P_ of the taproot internal key _p_, as defined in [BIP 341], to the stack."

Reply:
How about?
> When verifying taproot script path spends having leaf version `0xc0` (as defined in [BIP 342]), `OP_INTERNALKEY` replaces `OP_SUCCESS203` (0xcb). `OP_INTERNALKEY` pushes the 32-byte x-only representation (referred to as _P_) of the _taproot internal key_ (referred to as _p_), as defined in [BIP 341], to the stack.

Reply:
Note that BIP341 already defines the taproot internal key _p_ (note the lowercase) as a 32-byte array (as opposed to, say, a point on the secp256k1 curve), so it's not necessary to reinterpret it in any way before pushing it to the stack.

Reply:
If others feel that this passage is sufficiently clear, I don’t feel strongly about it.

Reply:
So then:

> When verifying taproot script path spends having leaf version `0xc0` (as defined in [BIP 342]), `OP_INTERNALKEY` replaces `OP_SUCCESS203` (0xcb). `OP_INTERNALKEY` pushes the 32-byte _taproot internal key_ (referred to as _p_), as defined in [BIP 341], to the stack.

?

Reply:
I find it weird that both _taproot internal key_ and _p_ are in italic and together like that. No hard preference just feels off somehow!

Reply:
I understand that, but it's not unprecedented, for example in BIP341 you'll find "_annex_ _a_". Also notice that they're in separate italic blocks, even though you can't tell from the rendered version.